.editorconfig

```diff
@@ -0,0 +1,191 @@
+# editorconfig.org
+
+# top-most EditorConfig file
+root = true
+
+# Default settings:
+# A newline ending every file
+# Use 4 spaces as indentation
+[*]
+insert_final_newline = true
+indent_style = space
+indent_size = 4
+trim_trailing_whitespace = true
+
+[project.json]
+indent_size = 2
+
+# C# files
+[*.cs]
+# New line preferences
+csharp_new_line_before_open_brace = all
+csharp_new_line_before_else = true
+csharp_new_line_before_catch = true
+csharp_new_line_before_finally = true
+csharp_new_line_before_members_in_object_initializers = true
+csharp_new_line_before_members_in_anonymous_types = true
+csharp_new_line_between_query_expression_clauses = true
+
+# Indentation preferences
+csharp_indent_block_contents = true
+csharp_indent_braces = false
+csharp_indent_case_contents = true
+csharp_indent_case_contents_when_block = true
+csharp_indent_switch_labels = true
+csharp_indent_labels = one_less_than_current
+
+# Modifier preferences
+csharp_preferred_modifier_order = public,private,protected,internal,static,extern,new,virtual,abstract,sealed,override,readonly,unsafe,volatile,async:suggestion
+
+# avoid this. unless absolutely necessary
+dotnet_style_qualification_for_field = false:suggestion
+dotnet_style_qualification_for_property = false:suggestion
+dotnet_style_qualification_for_method = false:suggestion
+dotnet_style_qualification_for_event = false:suggestion
+
+# Types: use keywords instead of BCL types, and permit var only when the type is clear
+csharp_style_var_for_built_in_types = false:suggestion
+csharp_style_var_when_type_is_apparent = false:none
+csharp_style_var_elsewhere = false:suggestion
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
+dotnet_style_predefined_type_for_member_access = true:suggestion
+
+# name all constant fields using PascalCase
+dotnet_naming_rule.constant_fields_should_be_pascal_case.severity = suggestion
+dotnet_naming_rule.constant_fields_should_be_pascal_case.symbols  = constant_fields
+dotnet_naming_rule.constant_fields_should_be_pascal_case.style    = pascal_case_style
+dotnet_naming_symbols.constant_fields.applicable_kinds   = field
+dotnet_naming_symbols.constant_fields.required_modifiers = const
+dotnet_naming_style.pascal_case_style.capitalization = pascal_case
+
+# static fields should have s_ prefix
+dotnet_naming_rule.static_fields_should_have_prefix.severity = suggestion
```

Comment:
This guidance comes from [here](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md)

In general it looks like we don't follow this, though. Do we want to keep the editor suggestion or not?

Reply:
How many warnings are you seeing? How far are we from the .net core guidance?

Reply:
Currently, "s_" prefix violations are at 144. But most of those are coming from opensource/json and test assemblies. 
After ignoring those two sources we are down to 48 I will write up some non root editorconfig files for those directories

Reply:
We should disable analyzers and code style formatting for opensource/json as we are not maintaining that code.

I'm debating if we should fix the violations and start enforcing the rules? 
@brentschmaltz - Do you have an option about this?

Reply:
@GeoK @keegan-caruso I don't think we should change any internal json code.

Reply:
fixed in a569748 Don't show any style warnings for tests or opensource/json.
Bringing the suggestions down to ~400.

Some of these would be good to fix (static prefix, don't qualify access with `this`) some of them are more in the realm of polish (simplifying object/collection initialization, use pattern matching, etc)